hotel/models/hotel_folio_room.py

```diff
@@ -0,0 +1,234 @@
+                self.folio_id.date_order,
+                False,
+                self.folio_id.partner_id
+                .property_account_position.id,
+                False,
+                self._context
+            )
+
+    @api.onchange('checkin_date', 'checkout_date')
+    def _onchange_dates(self):
+        if self.checkin_date and self.checkout_date:
+            duration = self._date_dif()
+            if self.product_uom_qty != duration:
+                self.product_uom_qty = duration
+            if self.checkin_date > self.checkout_date:
+                raise Warning(_('Check In can`t be higher than Check Out.'))
+
+    @api.one
+    @api.constrains('checkin_date', 'checkout_date', 'product_id')
+    def _check_room_dates(self):
+        if self.product_id.name:
+            room_id = self.product_id.id
+
+            rs = self.env['hotel.folio.room'].search_count([
+                    '|', '&',
+                    ('product_id', '=', room_id),
+                    '&',
+                    ('checkin_date', '<=', self.checkin_date),
+                    ('checkout_date', '>=', self.checkin_date),
+                    '|', '&',
+                    ('product_id', '=', room_id),
+                    '&',
+                    ('checkout_date', '>=', self.checkout_date),
+                    ('checkin_date', '<=', self.checkout_date),
+                    '&',
+                    ('product_id', '=', room_id),
+                    '&',
+                    ('checkin_date', '>=', self.checkin_date),
+                    ('checkout_date', '<=', self.checkout_date)])
+
+            if rs > 1:
+                raise ValidationError(_('Your selected rooms are \
+                 reserved for those \
+                 days. Please choose another dates.'))
+
+    @api.onchange('product_uom_qty')
+    def _onchange_duration(self):
+        if self.checkin_date and self.product_uom_qty:
+            checkin_date = datetime.datetime.strptime(
+                self.checkin_date,
+                DEFAULT_SERVER_DATETIME_FORMAT)
+            duration = datetime.timedelta(days=self.product_uom_qty)
+            checkout_date = checkin_date + duration
+            checkout_date = datetime.datetime.strftime(
+                checkout_date,
+                DEFAULT_SERVER_DATETIME_FORMAT)
+            if self.checkout_date != checkout_date:
+                self.checkout_date = checkout_date
+
+    def _date_dif(self):
+        checkin_date = datetime.datetime.strptime(
+                self.checkin_date,
+                DEFAULT_SERVER_DATETIME_FORMAT)
+        checkout_date = datetime.datetime.strptime(
+                self.checkout_date,
+                DEFAULT_SERVER_DATETIME_FORMAT)
+        duration_date = checkout_date - checkin_date
+        duration = duration_date.days
+        return duration
+
+    @api.multi
+    def on_change_checkout(self,
+                           checkin_date=time.strftime('%Y-%m-%d %H:%M:%S'),
```

Comment:
I think that you are going to have some unexpected behaviour if you do it like this.
The checkin_date and checkout_date should by default be equal to None.
Inside the function, you should check if they are none and then set them to time.strftime('%Y-%m-%d %H:%M:%S')